File: rules/core/testdata/safe/parameterized_query.ts
import express, { Request, Response } from "express";
import Knex from "knex";

const app = express();
const knex = Knex({ client: "pg", connection: "postgresql://localhost/mydb" });

app.get("/users/search", async (req: Request, res: Response) => {
  const name = req.query.name as string;

  // SAFE: knex.raw() with binding parameter
  const users = await knex.raw("SELECT * FROM users WHERE name = ?", [name]);
  res.json(users.rows);
});

app.get("/users/:id", async (req: Request, res: Response) => {
  const id = parseInt(req.params.id, 10);

  // SAFE: parseInt sanitizes input + parameterized query
  if (isNaN(id)) {
    res.status(400).json({ error: "invalid id" });
    return;
  }
  const user = await knex.raw("SELECT * FROM users WHERE id = ?", [id]);
  res.json(user.rows[0]);
});

File: rules/core/testdata/safe/parameterized_query.py
from flask import Flask, request, jsonify
from sqlalchemy import create_engine, text

app = Flask(__name__)
engine = create_engine("postgresql://localhost/mydb")


@app.route("/users/search")
def search_users():
    name = request.args.get("name")

    # SAFE: bound parameters via bindparams
    with engine.connect() as conn:
        result = conn.execute(
            text("SELECT id, email FROM users WHERE name = :name").bindparams(name=name)
        )
        rows = [dict(row._mapping) for row in result]
    return jsonify(rows)


@app.route("/users/<user_id>/notes")
def get_notes(user_id):
    search = request.args.get("q", "")

    # SAFE: parameterized query with named placeholders
    uid = int(user_id)
    with engine.connect() as conn:
        result = conn.execute(
            text("SELECT * FROM notes WHERE user_id = :uid AND body LIKE :search"),
            {"uid": uid, "search": f"%{search}%"},
        )
        rows = [dict(row._mapping) for row in result]
    return jsonify(rows)
